hutch_python/ipython_session_timer.py

```diff
@@ -0,0 +1,94 @@
+"""
+This module modifies an ``ipython`` shell to automatically close if it has been
+idle for 48 hours. Prior to starting hutch-python the user is alerted that the
+shell will automatically close after another 24 hours and is given the option
+to enter the number of hours they want to extend the session.
+"""
+
+import time
+from threading import Thread
+
+from IPython import get_ipython
+
+max_idle_time = 172800  # number of seconds in 48 hours
+
+
+def configure_timeout(session_timer):
+    global max_idle_time
+    if isinstance(session_timer, int) and session_timer > 0:
+        max_idle_time = session_timer
+
+
+class IPythonSessionTimer:
+    '''
+    Class tracks the amount of time the current `InteractiveShell` instance (henceforth
+    called 'user session') has been idle and closes the session if more than 48
+    hours have passed.
+
+    Time is in seconds (floating point) since the epoch began. (In UNIX the
+    epoch started on January 1, 1970, 00:00:00 UTC)
+
+    Parameters
+    ----------
+    ipython : ``IPython.terminal.interactiveshell.TerminalInteractiveShell``
+        The active ``ipython`` ``Shell``, perhaps the one returned by
+        ``IPython.get_ipython()``.
+
+    Attributes
+    ----------
+    max_idle_time: int
+        The maximum number of seconds a user session can be idle (currently set
+        to 172800 seconds or 48 hours).
+    '''
+
+    def __init__(self, ipython):
+        global max_idle_time
+        self.max_idle_time = max_idle_time
+        self.user_active = False
+
+        # _set_last_active_time() function will trigger every time user runs a cell
+        ipython.events.register('pre_run_cell', self._set_user_active)
+        ipython.events.register('post_run_cell', self._set_user_inactive)
+
+    def _set_user_active(self, result):
+        self.user_active = True
+
+    def _set_user_inactive(self, result):
+        self.user_active = False
+
+    def _start_session(self):
+        # poll for user activity
+        while (1):
+            if not self.user_active:
+                time.sleep(self.max_idle_time)
+                if not self.user_active:
```

Comment:
This logic works for long running cells, but doesn't "refresh" the time when a cell is quick.  Quick cells flip the active flag back and forth to false immediately, so the idle sleep timer from the previous cell can break this loop.

You can test this by running some long cell, then repeatedly doing some simple assignment.  

I think we actually need to keep track of the end time of the last completed cell as well, and timeout based on that as well.

Reply:
I couldn't get the implementation using `pre_run_cell` and `post_run_cell` to cancel the threading.Timer() function. There is a stackoverflow post about using a global t = Timer() function but doing this with self.T = Timer() didn't work (https://stackoverflow.com/questions/56566801/python-timer-start-and-reset). I went back to the first iteration of ipython_session_timer.py and tried to create a combination that works. I think the latest commit should be working (hope nothing breaks!)  Will try to add some unit tests and fix other issues.